### PR TITLE
Add ViT finetuning post

### PR DESCRIPTION
**Pull request by nateraw**
A blogpost on how to fine-tune a Vision transformer with `transformers.Trainer` and `datasets`. I think it would be really great to get this one in soon, and then follow it up with #226 which builds on the ideas here.

### CODE REVIEW

File: fine-tune-vit.md [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+
+
+    {
+      'pixel_values': tensor([[[[-0.6157, -0.6000, -0.6078,  ..., ]]]]),
+      'labels': 0
+    }
+
+
+
+While we could call `ds.map` and apply this to every example at once, this can be very slow, especially if you use a larger dataset. Instead, we'll apply a ***transform*** to the dataset. Transforms are only applied to examples as you index them.
```
Comment (osanseviero) [inline fine-tune-vit.md:269]:
```suggestion
While you could call `ds.map` and apply this to every example at once, this can be very slow, especially if you use a larger dataset. Instead, you can apply a ***transform*** to the dataset. Transforms are only applied to examples as you index them.
```

File: fine-tune-vit.md [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+
+First, though, we'll need to update our last function to accept a batch of data, as that's what `ds.with_transform` expects.
```
Comment (osanseviero) [inline fine-tune-vit.md:271]:
```suggestion
First, though, you'll need to update the last function to accept a batch of data, as that's what `ds.with_transform` expects.
```

File: fine-tune-vit.md [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+def transform(example_batch):
+    # Take a list of PIL images and turn them to pixel values
+    inputs = feature_extractor([x for x in example_batch['image']], return_tensors='pt')
+
+    # Don't forget to include the labels!
+    inputs['labels'] = example_batch['labels']
+    return inputs
+```
+
+We can directly apply this to our dataset using `ds.with_transform(transform)`.
```
Comment (osanseviero) [inline fine-tune-vit.md:286]:
```suggestion
You can directly apply this to the dataset using `ds.with_transform(transform)`.
```

File: fine-tune-vit.md [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+
+
+```python
+prepared_ds = ds.with_transform(transform)
+```
+
+Now, whenever we get an example from the dataset, our transform will be
```
Comment (osanseviero) [inline fine-tune-vit.md:293]:
```suggestion
Now, whenever you get an example from the dataset, the transform will be 
```

File: fine-tune-vit.md [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+    {
+      'pixel_values': tensor([[[[-0.6157, -0.6000, -0.6078,  ..., ]]]]),
+      'labels': [0, 0]
+    }
+
+
+
+# Training and Evaluation
+
+The data is processed and we are ready to start setting up the training pipeline. We will make use of 🤗's Trainer, but that'll require us to do a few things first:
```
Comment (osanseviero) [inline fine-tune-vit.md:313]:
```suggestion
The data is processed and you are ready to start setting up the training pipeline. This blog post uses 🤗's Trainer, but that'll require to do a few things first:
```

File: fine-tune-vit.md [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+
+- Define a collate function.
+
+- Define an evaluation metric. During training, the model should be evaluated on its prediction accuracy. We should define a compute_metrics function accordingly.
```
Comment (osanseviero) [inline fine-tune-vit.md:317]:
```suggestion
- Define an evaluation metric. During training, the model should be evaluated on its prediction accuracy. You should define a `compute_metrics` function accordingly.
```

File: fine-tune-vit.md [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+
+- Load a pretrained checkpoint. We need to load a pretrained checkpoint and configure it correctly for training.
+
+- Define the training configuration.
+
+After having fine-tuned the model, we will correctly evaluate it on the evaluation data and verify that it has indeed learned to correctly classify our images.
```
Comment (osanseviero) [inline fine-tune-vit.md:323]:
```suggestion
After fine-tuning the model, you will correctly evaluate it on the evaluation data and verify that it has indeed learned to correctly classify our images.
```

File: fine-tune-vit.md [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+def collate_fn(batch):
+    return {
+        'pixel_values': torch.stack([x['pixel_values'] for x in batch]),
+        'labels': torch.tensor([x['labels'] for x in batch])
+    }
+```
+
+### Define an evaluation metric
+
+Here, we load the [accuracy](https://huggingface.co/metrics/accuracy) metric from `datasets`, and then write a function that takes in a model prediction + computes the accuracy.
```
Comment (osanseviero) [inline fine-tune-vit.md:344]:
```suggestion
The [accuracy](https://huggingface.co/metrics/accuracy) metric from `datasets` can easily be used to compare the predictions with the labels.
```

File: fine-tune-vit.md [hunk 9]
```diff
@@ -0,0 +1,467 @@
...
+import numpy as np
+from datasets import load_metric
+
+metric = load_metric("accuracy")
+def compute_metrics(p):
+    return metric.compute(predictions=np.argmax(p.predictions, axis=1), references=p.label_ids)
+```
+
+
+Now we can load our pretrained model. We'll add `num_labels` on init to make sure the model creates a classification head with the right number of units. We'll also include the `id2label` and `label2id` mappings so we have human readable labels in the 🤗 hub widget if we choose to `push_to_hub`.
```
Comment (osanseviero) [inline fine-tune-vit.md:357]:
```suggestion
Let's load the pretrained model. We'll add `num_labels` on init so the model creates a classification head with the right number of units. We'll also include the `id2label` and `label2id` mappings to have human-readable labels in the Hub widget.
```

File: fine-tune-vit.md [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+model = ViTForImageClassification.from_pretrained(
+    model_name_or_path,
+    num_labels=len(labels),
+    id2label={str(i): c for i, c in enumerate(labels)},
+    label2id={c: str(i) for i, c in enumerate(labels)}
+)
+```
+
+
+We're almost ready to train! The last thing we'll do before that is set up the training configuration by defining [`TrainingArguments`](https://huggingface.co/docs/transformers/v4.16.2/en/main_classes/trainer#transformers.TrainingArguments).
```
Comment (osanseviero) [inline fine-tune-vit.md:374]:
```suggestion
Almost ready to train! The last thing needed before that is to set up the training configuration by defining [`TrainingArguments`](https://huggingface.co/docs/transformers/v4.16.2/en/main_classes/trainer#transformers.TrainingArguments).
```

File: fine-tune-vit.md [hunk 11]
```diff
@@ -0,0 +1,467 @@
...
+
+Most of these are pretty self-explanatory, but one that is quite important here is `remove_unused_columns=False`. This one will drop any features not used by the model's call function. By default it's `True` because usually its ideal to drop unused feature columns, as it makes it easier to unpack inputs into the model's call function. But, in our case, we need the unused features ('image' in particular) in order to create 'pixel_values'.
+
+What I'm trying to say is that you'll have a bad time if you forget to set `remove_unused_columns=False`.
```
Comment (osanseviero) [inline fine-tune-vit.md:378]: 🤣

File: fine-tune-vit.md [hunk 12]
```diff
@@ -0,0 +1,467 @@
...
+    ***** eval metrics *****
+      epoch                   =        4.0
+      eval_accuracy           =      0.985
+      eval_loss               =     0.0637
+      eval_runtime            = 0:00:02.13
+      eval_samples_per_second =     62.356
+      eval_steps_per_second   =       7.97
+
+
+Finally, if you want, you can push your model up to the hub. Here, we'll push it up if you specified `push_to_hub=True` in the training configuration.
```
Comment (osanseviero) [inline fine-tune-vit.md:451]:
```suggestion
Finally, if you want, you can push your model to the Hub. Here, we'll push it up if you specified `push_to_hub=True` in the training configuration.
```

File: fine-tune-vit.md [hunk 13]
```diff
@@ -0,0 +1,467 @@
...
+    "tags": ['image-classification'],
+}
+
+if training_args.push_to_hub:
+    trainer.push_to_hub('🍻 cheers', **kwargs)
+else:
+    trainer.create_model_card(**kwargs)
+```
+
+The resulting model has been shared to [nateraw/vit-base-beans](https://huggingface.co/nateraw/vit-base-beans). I'm assuming you don't have pictures of bean leaves laying around, but if you do, give it a try! 🚀
```
Comment (osanseviero) [inline fine-tune-vit.md:467]: You could add some example images with https://huggingface.co/docs/hub/model-repos#how-can-i-control-my-models-widget-example-input

**Review by osanseviero — COMMENTED** [review]
Very cool post! There are many comments but most of them are nits and small gramma changes. Some quick things
* We should compress the images
* Consider adding a embedded gradio demo at the end if possible or adding examples to the widget (last comment)
* Avoid using first person. The switch between first and third person is a bit strange

Again, the blog post is pretty much ready imo and looks almost ready to be shipped 🚢

File: fine-tune-vit.md [hunk 14]
```diff
@@ -0,0 +1,467 @@
...
+</a>
+
+Just as transformers-based models have revolutionized NLP, we're now seeing an explosion of papers applying them to all sorts of other domains. One of the most revolutionary of these was the Vision Transformer (ViT), which was introduced in [June 2021](https://arxiv.org/abs/2010.11929) by a team of researchers at Google Brain.
+
+This paper explored how you can tokenize images, just as you would tokenize sentences, so that they can be passed to transformer models for training. Its quite a simple concept, really...
+
+1. Split an image into a grid of sub-image patches
+1. Embed each patch with a linear projection
+1. Each embedded patch becomes a token, and the resulting sequence of embedded patches is the sequence you pass to the model.
+
```
[the remark this answers is not in the document]
Comment (nateraw) [reply fine-tune-vit.md:40]: Fun fact, in Markdown you can use a sequence of numbers that don't increase, and markdown will render them correctly for you. Here's what it looks like in moon landing:

---

Screen Shot 2022-02-10 at 11 33 11 AM

File: fine-tune-vit.md [hunk 15]
```diff
@@ -0,0 +1,467 @@
...
+![vit_figure.png](https://raw.githubusercontent.com/google-research/vision_transformer/main/vit_figure.png)
```
[the remark this answers is not in the document]
Comment (nateraw) [reply fine-tune-vit.md:41]: Yes - and a note for myself...this should be a local image not the URL. I downloaded it, so it should be opening directly instead.

File: fine-tune-vit.md [hunk 16]
```diff
@@ -0,0 +1,467 @@
...
+
+
+```python
+from datasets import load_dataset
+
+ds = load_dataset('beans')
+ds
+```
+
+Let's take a look at the 400th example from the `'train'` split from the beans dataset. You'll notice each example from the dataset has 3 features:
```
[the remark this answers is not in the document]
Comment (nateraw) [reply fine-tune-vit.md:71]: Disagree on this one - more of a personal voice thing here.

File: fine-tune-vit.md [hunk 17]
```diff
@@ -0,0 +1,467 @@
...
+
+1. `image`: A PIL Image
+1. `image_file_path`: The `str` path to the image file that was loaded as `image`
+1. `labels`: A [`datasets.ClassLabel`](https://huggingface.co/docs/datasets/package_reference/main_classes.html?highlight=classlabel#datasets.ClassLabel) feature, which we'll see as an integer representation of the label for a given example. (Later we'll see how to get the string class names, don't worry)
```
[the remark this answers is not in the document]
Comment (nateraw) [reply fine-tune-vit.md:75]: hmm, I see what you're trying to change here. I'll edit this one myself. Will mark as resolved when I do so.

File: fine-tune-vit.md [hunk 18]
```diff
@@ -0,0 +1,467 @@
...
+
+When ViT models are trained, specific transformations are applied to images being fed into them. Use the wrong transformations on your image and the model won't be able to understand what it's seeing! 🖼 ➡️ 🔢
+
+To make sure we apply the correct transformations, we will use a [`ViTFeatureExtractor`](https://huggingface.co/docs/datasets/package_reference/main_classes.html?highlight=classlabel#datasets.ClassLabel.int2str) initialized with a configuration that was saved along with the pretrained model we plan to use. In our case, we'll be using the [google/vit-base-patch16-224-in21k](https://huggingface.co/google/vit-base-patch16-224-in21k) model, so lets load its feature extractor from the 🤗 Hub.
+
+
+```python
+from transformers import ViTFeatureExtractor
+
+model_name_or_path = 'google/vit-base-patch16-224-in21k'
```
[the remark this answers is not in the document]
Comment (nateraw) [reply fine-tune-vit.md:199]: I think I'll leave this one as-is. Somewhat of a personal preference 🤷
